# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1235717480 -3600
#      Fri Feb 27 07:51:20 2009 +0100
# Node ID 8633ae24a89231f5d379a9e44fd9df74eab1bef2
# Parent  8ed42c679af5ddc224865fd6b2cee03111b5e204
gripe when dividing a diagonal matrix by zero

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,14 @@
+2009-02-27  Jaroslav Hajek  <highegg@gmail.com>
+
+	* OPERATORS/op-dms-template.cc (gripe_if_zero): New template static
+	function.
+	(dmsdiv, sdmldiv): Call it.
+
 2009-02-26  John W. Eaton  <jwe@octave.org>
 
 	* symtab.h (symbol_table::symbol_record::symbol_record_rep::forced):
 	New static constant.
 	(symbol_table::symbol_record::symbol_record_rep::force_variable,
 	symbol_table::symbol_record::force_variable): New functions.
 	(symbol_table::symbol_record::symbol_record_rep::is_forced,
 	symbol_table::symbol_record::symbol_record_rep::mark_forced,
diff --git a/src/OPERATORS/op-dms-template.cc b/src/OPERATORS/op-dms-template.cc
--- a/src/OPERATORS/op-dms-template.cc
+++ b/src/OPERATORS/op-dms-template.cc
@@ -20,16 +20,17 @@ along with Octave; see the file COPYING.
 
 */
 
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
 
 #include "ops.h"
+#include "gripes.h"
 #include "xpow.h"
 #include SINCLUDE
 #include MINCLUDE
 
 // matrix by diag matrix ops.
 
 #ifndef SCALARV
 #define SCALARV SCALAR
@@ -40,28 +41,43 @@ along with Octave; see the file COPYING.
 #endif
 
 DEFNDBINOP_OP (sdmadd, SCALAR, MATRIX, SCALARV, MATRIXV, +)
 DEFNDBINOP_OP (sdmsub, SCALAR, MATRIX, SCALARV, MATRIXV, -)
 DEFNDBINOP_OP (sdmmul, SCALAR, MATRIX, SCALARV, MATRIXV, *)
 DEFNDBINOP_OP (dmsadd, MATRIX, SCALAR, MATRIXV, SCALARV, +)
 DEFNDBINOP_OP (dmssub, MATRIX, SCALAR, MATRIXV, SCALARV, -)
 DEFNDBINOP_OP (dmsmul, MATRIX, SCALAR, MATRIXV, SCALARV, *)
-DEFNDBINOP_OP (dmsdiv, MATRIX, SCALAR, MATRIXV, SCALARV, /)
 
 #define OCTAVE_MATRIX CONCAT2(octave_, MATRIX)
 #define OCTAVE_SCALAR CONCAT2(octave_, SCALAR)
 #define MATRIX_VALUE CONCAT2(MATRIXV, _value)
 #define SCALAR_VALUE CONCAT2(SCALARV, _value)
 
+template <class T>
+static T 
+gripe_if_zero (T x)
+{
+  if (x == T ())
+    gripe_divide_by_zero ();
+  return x;
+}
+
+DEFBINOP (dmsdiv, MATRIX, SCALAR)
+{
+  CAST_BINOP_ARGS (const OCTAVE_MATRIX&, const OCTAVE_SCALAR&);
+
+  return v1.MATRIX_VALUE () / gripe_if_zero (v2.SCALAR_VALUE ());
+}
+
 DEFBINOP (sdmldiv, SCALAR, MATRIX)
 {
   CAST_BINOP_ARGS (const OCTAVE_SCALAR&, const OCTAVE_MATRIX&);
 
-  return v2.MATRIX_VALUE () / v1.SCALAR_VALUE ();
+  return v2.MATRIX_VALUE () / gripe_if_zero (v1.SCALAR_VALUE ());
 }
 
 DEFBINOP (dmspow, MATRIX, SCALAR)
 {
   CAST_BINOP_ARGS (const OCTAVE_MATRIX&, const OCTAVE_SCALAR&);
 
   return xpow (v1.MATRIX_VALUE (), v2.SCALAR_VALUE ());
 }
